include/AdePT/kernels/gammas.cuh

```diff
@@ -77,20 +77,36 @@ __global__ void TransportGammas(adept::TrackManager<Track> *gammas, Secondaries
     theTrack->SetMCIndex(auxData.fMCIndex);
 
     // Sample the `number-of-interaction-left` and put it into the track.
-    for (int ip = 0; ip < 3; ++ip) {
-      double numIALeft = currentTrack.numIALeft[ip];
-      if (numIALeft <= 0) {
-        numIALeft = -std::log(currentTrack.Uniform());
-      }
-      theTrack->SetNumIALeft(numIALeft, ip);
+    // Use index 0 since numIALeft for gammas is based only on the total macroscopic cross section
+    if (theTrack->GetNumIALeft(0) <= 0.0) {
+      theTrack->SetNumIALeft(-std::log(currentTrack.Uniform()), 0);
     }
 
     // Call G4HepEm to compute the physics step limit.
     G4HepEmGammaManager::HowFar(&g4HepEmData, &g4HepEmPars, &gammaTrack);
+    G4HepEmGammaManager::SampleInteraction(&g4HepEmData, &gammaTrack, currentTrack.Uniform());
+    int winnerProcessIndex = theTrack->GetWinnerProcessIndex();
+
+    // avoid photo-nuclear reaction that would need to be handled by G4 itself
+    if (winnerProcessIndex == 3) {
+      winnerProcessIndex = -1;
+      // NOTE: re-drawing was found to change the physics slightly, thus, it is commented out for now
+      // // since we do an redundant step if gamma-nuclear is drawn, we redraw up to 3 times
+      // int trials_left = 3;
+      // do {
+      //   // reset `number-of-interaction-left` since it is always consumed by the previous call to SampleInteraction()
+      //   theTrack->SetNumIALeft(-std::log(currentTrack.Uniform()), 0);
```

Comment:
Shouldn't this rather be
```suggestion
      //   theTrack->SetNumIALeft(-std::log(currentTrack.Uniform()), 3);
```
?

Reply:
Thanks for checking! I think for gammas now only the 0 index should be used, which was introduced in [this PR](https://github.com/mnovak42/g4hepem/pull/115). Easier to find in [these lines](https://github.com/mnovak42/g4hepem/blob/master/G4HepEm/G4HepEmRun/include/G4HepEmGammaManager.icc#L93-L98).

Reply:
Ah ok. But then I think it doesn't need to be set at all because `SampleInteraction` just resets it...

Also, is this approach correct? It's resampling an interaction from a point that was reached with a mfp including the gamma-nuclear interactions, but without actually moving. I guess this is why you are seeing slight changes in physics...

Reply:
Just for documentation:

After an offline discussion, we agreed that no simple re-drawing is possible. The `HowFar` method returns now smaller steps due to the gamma-nuclear reactions in comparison to without gamma-nuclear reactions. If we then just re-draw we get shorter steps which is exactly what was observed in form of a lower energy deposition in the tail region for testEm3 and a higher in the front. Thus, we must do an empty step without a reaction to compensate for the smaller step size returned by `HowFar`.

Additional note: the `SampleInteractions` only needs to be called after the geometry step limit is known, since we don't need to sample the interaction if we are limited by geometry anyway.